Support cellular IPv4v6 dual stack in LWIP

mikaleppanen opened the pull request:
## Description
Cellular IPv4v6 dual stack support to lwip.

- Merged ipv6 address auto negotiation corrections from lwip master branch, which prevented ipv6 cellular connection from working (lwip did not work correctly, if RA prefix had auto negotiation flag set, but on-link flag was not set).
- Made correction to lwip to disable ipv6 neighbor solidification probes, if link layer address is not available. 
- Disabled duplicated address detection for cellular ipv6. 
- If cellular dual stack support is enabled, tries first cellular context activation for both ipv4v6, and if context activation is not successful, falls back to ipv4 context.
- If context is activated for ipv4v6, makes PPP negotiation with both IPCP and IPV6CP protocols, and configures both ipv4 and ipv6 addresses.
- If context is only for ipv4, makes PPP negotiation with IPCP protocol, and configures ipv4 address.
- If both ipv4 and ipv6 addresses are configured, when DNS address resolution is made, tries to resolve first the preferred stacks address. If address resolution fails, tries to resolve non-preferred stack address.
- Added trace dump function that can be used to trace PPP, ipv4 and ipv6 packets in format that can be imported to Wireshark.
- Moved lwip memory pools to second memory bank for UBLOX_C027.
- Corrected lwip random and tcp isn not to include configuration headers.

## Status
READY


## Migrations
YES

nsapi_ppp.h function nsapi_ppp_connect(): Added "const nsapi_ip_stack_t stack" parameter
-> can be used in dual stack case to set either ipv4v6 or ipv4 stack, for non-dual
stack case use value DEFAULT_STACK.

ppp_lwip.h function ppp_lwip_if_init(): Added "const nsapi_ip_stack_t stack" parameter
-> can be used in dual stack case to set either ipv4v6 or ipv4 stack, for non-dual
stack case use value DEFAULT_STACK.

lwip_stack.h function mbed_lwip_bringup_2(): Added "const nsapi_ip_stack_t stack" parameter
-> can be used in dual stack case to set either ipv4v6 or ipv4 stack, for non-dual
stack case use value DEFAULT_STACK.

## Related PRs
None


## Todos
- [ ] Tests
- [ ] Documentation


## Deploy notes
None


## Steps to test or reproduce
None

mikaleppanen commented:
@SeppoTakalo @kjbracey-arm @hasnainvirk please review.

hasnainvirk commented on features/FEATURE_LWIP/lwip-interface/lwip_stack.c:
```diff
@@ -547,43 +547,49 @@ nsapi_error_t mbed_lwip_bringup_2(bool dhcp, bool ppp, const char *ip, const cha
-    ip6_addr_set_allnodes_linklocal(&ip6_allnodes_ll);
-    lwip_netif.mld_mac_filter(&lwip_netif, &ip6_allnodes_ll, NETIF_ADD_MAC_FILTER);
-  }
+        /*
```
Formatting

mikaleppanen replied:
just moved the indentation when if was added, no problem there.

hasnainvirk commented on features/netsocket/cellular/generic_modem_driver/PPPCellularInterface.cpp:
```diff
@@ -262,6 +262,7 @@ PPPCellularInterface::PPPCellularInterface(FileHandle *fh, bool debug)
     _pwd = NULL;
     _fh = fh;
     _debug_trace_on = debug;
+    _stack = IPV4_STACK;
```
Shouldn't this be IPV4V6_STACK ?

mikaleppanen replied:
I set the value to default stack. In that case flag configuration defines the stack.

hasnainvirk commented on features/netsocket/nsapi_ppp.h:
```diff
@@ -36,10 +36,11 @@ NetworkStack *nsapi_ppp_get_stack();
  *  @return             0 on success, negative error code on failure
  */
-nsapi_error_t nsapi_ppp_connect(FileHandle *stream, Callback<void(nsapi_error_t)> status_cb=0, const char *uname=0, const char *pwd=0);
+nsapi_error_t nsapi_ppp_connect(FileHandle *stream, Callback<void(nsapi_error_t)> status_cb=0, const char *uname=0, const char *pwd=0, const nsapi_ip_stack_t stack=IPV4_STACK);
```
Why default to IPv4 ? Isn't the whole point to have the stack dual configurable and choose at runtime which one is available ?

mikaleppanen replied:
Set also this to default stack.

hasnainvirk commented on features/FEATURE_LWIP/lwip-interface/lwipopts.h:
```diff
@@ -278,20 +283,25 @@
-#define PRINTPKT_SUPPORT            0
+#define PAP_SUPPORT                    0
+#define VJ_SUPPORT                     0
+#define PRINTPKT_SUPPORT               0
```
Formatting

mikaleppanen replied:
moved the indentation when longer define was added, no problem there.

hasnainvirk commented on features/FEATURE_LWIP/lwip-interface/lwipopts.h:
```diff
@@ -278,20 +283,25 @@
-#define IP_SOF_BROADCAST            0
-#define IP_SOF_BROADCAST_RECV       0
+#define IP_SOF_BROADCAST               0
+#define IP_SOF_BROADCAST_RECV          0
```
formating

hasnainvirk commented on features/FEATURE_LWIP/lwip-interface/lwipopts.h:
```diff
@@ -278,20 +283,25 @@
-#define MAXNAMELEN                  64     /* max length of hostname or name for auth */
-#define MAXSECRETLEN                64
+#define MAXNAMELEN                     64     /* max length of hostname or name for auth */
+#define MAXSECRETLEN                   64
```
Formatting

hasnainvirk commented on features/FEATURE_LWIP/lwip-interface/mbed_lib.json:
```diff
@@ -16,7 +16,11 @@
         "addr-timeout": {
```
Is this left out for backward compatibility ? Why and how ? Isn't the lwipopts.h also updated ?

mikaleppanen replied:
"addr-timeout" is the amount of time we wait for preferred system address in case we have only non-preferred system address available.
"both-addr-timeout" is the time to wait that both systems addresses are available. 

So the flags have different meaning and both are valid.

hasnainvirk commented on features/FEATURE_LWIP/lwip-interface/lwipopts.h:
```diff
@@ -278,20 +283,25 @@
 
-#define MAXNAMELEN                  64     /* max length of hostname or name for auth */
-#define MAXSECRETLEN                64
+#define MAXNAMELEN                     64     /* max length of hostname or name for auth */
```
Formatting

hasnainvirk commented on features/FEATURE_LWIP/lwip-interface/lwipopts.h:
```diff
@@ -278,20 +283,25 @@
-#define PAP_SUPPORT                 0
-#define VJ_SUPPORT                  0
-#define PRINTPKT_SUPPORT            0
+#define PAP_SUPPORT                    0
```
Formatting

mikaleppanen commented:
"Hasnain: there will be issues while using dual stack config, as 'add_dns_addr()". 

Yes, this could be improved maybe so that in dual stack system the DNS resolution could fallback to non-preferred system if it does not for some reason work in more preferred. I'll test this a little to see
how it works. 

There is no absolute need to modify the function since the IP address get function (mbed_lwip_get_ip_addr()) fetches either the preferred systems IP address or if not available the non-preferred. So a either ipv4 or ipv6 DNS server (google) will be always added to list if network has not provided DNS server.

mikaleppanen commented:
Jenkins is failing in interface up/down/up/down test since variable holding the address status is not reset on bring down. I'll correct that.

mikaleppanen commented:
Review defects corrected. 

I improved the DNS address addition functionality in case network does not provide DNS server addresses. Now in dual mode both ipv4 and ipv6 DNS addresses are added in order of preference defined by flag configuration.

Corrected IAR memory pools relocation for UBLOX_C027. 

Corrected interface up/down/up/down error detected by Jenkins.

hasnainvirk commented:
@mikaleppanen I like the changes you did to place some lwip stuff on .ethusbram section for IAR compiler. Great work. 
Can you please check why CI is failing ? Is this related to RAAS unavailability or a genuine case ?

mikaleppanen commented:
Yes, ble raas is down. Tools team is aware about it.

mikaleppanen commented:
Updated lwip not to add neighbor cache entry based on router advertisement if link layer address is not present (PPP case)

kjbracey-arm commented on features/FEATURE_LWIP/lwip-interface/lwip/src/netif/ppp/lwip_auth.c:
```diff
@@ -751,6 +751,10 @@ void link_established(ppp_pcb *pcb) {
     if (!doing_multilink) {
 	for (i = 0; (protp = protocols[i]) != NULL; ++i)
 	    if (protp->protocol != PPP_LCP
+#if PPP_IPV4_SUPPORT && PPP_IPV6_SUPPORT
```
This feels like it's in the wrong layer. Shouldn't a suppression check be in ip[v6]cp_lowerup?

mikaleppanen replied:
Corrected.

kjbracey-arm commented on features/FEATURE_LWIP/lwip-interface/lwip/src/core/ipv6/lwip_nd6.c:
```diff
@@ -1736,9 +1737,14 @@ nd6_new_router(const ip6_addr_t *router_addr, struct netif *netif)
-    neighbor_cache[neighbor_index].state = ND6_INCOMPLETE;
-    neighbor_cache[neighbor_index].counter.probes_sent = 1;
-    nd6_send_neighbor_cache_probe(&neighbor_cache[neighbor_index], ND6_SEND_FLAG_MULTICAST_DEST);
+    if (netif->hwaddr_len) {
```
Suggested simple alternative - set state to STALE if interface has no HW address. I believe that should stay in STALE forever, without triggering probes. (Because the entries are not used, so don't transition to DELAY)

mikaleppanen replied:
Corrected.

kjbracey-arm commented on features/FEATURE_LWIP/lwip-interface/mbed_lib.json:
```diff
@@ -16,7 +16,11 @@
             "value": 5
-        },
+        },        
+        "both-addr-timeout": {
```
Maybe comment to say that it's not terribly meaningful to set both options. How about a boolean that changes what addr-timeout means - extra wait for preferred, or extra wait for all addresses. Make the help clear that is the extra wait on top of the initial wait for any address.

mikaleppanen replied:
Corrected.

mikaleppanen commented:
@kjbracey-arm corrected defects.

mikaleppanen commented:
I think all has approved this. @hasnainvirk can you update status review status. @theotherjimmy @0xc0170 can this progress?

0xc0170 commented:
/morph test-nightly

0xc0170 commented:
@mikaleppanen How this can be tested ? Is it referenced PR above to cliapp? Any new example (will cellular example get an update) ?

mikaleppanen commented:
I have tested this manually using cliapp tests (above pr) and mbed-os-example-application with MTS_DRAGONFLY_F411RE. Cellular example will be updated but currently we do not have echo server for ipv6 connectivity (ublox server works only for ipv4).

mikaleppanen commented:
Updated pull request. Changed memory pool areas that caused nightly tests to fail. Tested to compile with failing targets and tested Ethernet connectivity with LPC1768 and UBLOX_C027.

SeppoTakalo commented:
@0xc0170 Please restart the test.

mikaleppanen commented:
Could you retest this.

SeppoTakalo commented:
@0xc0170 @adbridge Please, somebody restart the tests.

studavekar commented:
/morph test-nightly
